src/audio/audio_stream.c

```diff
@@ -0,0 +1,175 @@
+					      struct sof_ipc_stream_params *params,
+					      bool force_update)
+{
+	struct audio_stream __sparse_cache *audio_stream =
+			attr_container_of(source, struct audio_stream __sparse_cache,
+					  source_api, __sparse_cache);
+	struct comp_buffer __sparse_cache *buffer =
+			attr_container_of(audio_stream, struct comp_buffer __sparse_cache,
+					  stream, __sparse_cache);
+
+	return buffer_set_params(buffer, params, force_update);
+}
+
+static int audio_stream_set_ipc_params_sink(struct sof_sink __sparse_cache *sink,
+					    struct sof_ipc_stream_params *params,
+					    bool force_update)
+{
+	struct audio_stream __sparse_cache *audio_stream =
+			attr_container_of(sink, struct audio_stream __sparse_cache,
+					  sink_api, __sparse_cache);
+	struct comp_buffer __sparse_cache *buffer =
+			attr_container_of(audio_stream, struct comp_buffer __sparse_cache,
+					  stream, __sparse_cache);
+
+	return buffer_set_params(buffer, params, force_update);
+}
+
+static int audio_stream_source_set_alignment_constants(struct sof_source __sparse_cache *source,
+						       const uint32_t byte_align,
+						       const uint32_t frame_align_req)
+{
+	struct audio_stream __sparse_cache *audio_stream =
+			attr_container_of(source, struct audio_stream __sparse_cache,
+					  source_api, __sparse_cache);
+
+	audio_stream_init_alignment_constants(byte_align, frame_align_req, audio_stream);
+
+	return 0;
+}
+
+static int audio_stream_sink_set_alignment_constants(struct sof_sink __sparse_cache *sink,
+						     const uint32_t byte_align,
+						     const uint32_t frame_align_req)
+{
+	struct audio_stream __sparse_cache *audio_stream =
+			attr_container_of(sink, struct audio_stream __sparse_cache,
+					  sink_api, __sparse_cache);
+
+	audio_stream_init_alignment_constants(byte_align, frame_align_req, audio_stream);
+
+	return 0;
+}
+
+static const struct source_ops audio_stream_source_ops = {
+	.get_data_available = audio_stream_get_data_available,
+	.get_data = audio_stream_get_data,
+	.release_data = audio_stream_release_data,
+	.audio_set_ipc_params = audio_stream_set_ipc_params_source,
+	.set_alignment_constants = audio_stream_source_set_alignment_constants
+};
+
+static const struct sink_ops audio_stream_sink_ops = {
+	.get_free_size = audio_stream_get_free_size,
+	.get_buffer = audio_stream_get_buffer,
+	.commit_buffer = audio_stream_commit_buffer,
+	.audio_set_ipc_params = audio_stream_set_ipc_params_sink,
+	.set_alignment_constants = audio_stream_sink_set_alignment_constants
+};
+
+void audio_stream_init(struct audio_stream __sparse_cache *audio_stream,
+		       void *buff_addr, uint32_t size)
+{
+	audio_stream->size = size;
+	audio_stream->addr = buff_addr;
+	audio_stream->end_addr = (char *)audio_stream->addr + size;
+
+	/* set the default alignment info.
+	 * set byte_align as 1 means no alignment limit on byte.
+	 * set frame_align as 1 means no alignment limit on frame.
```

Comment:
no idea what this means either.

Reply:
there is a limited calculation frame number function audio_stream_avail_frames_aligned, it use shift to replace division, the byte_align and frame_align is for shift index calculation.